README.md
# devcenter-backend

A backend to handle everything game/developer related.

## API

The service must be accessed by HTTPS.

### Player Facing

#### List all games

##### Request

**GET** ``/public/games``

###### Body

Empty.

##### Response

###### Body

JSON encoded array of games like this:

```javascript
{
  "games": [
    {
      "uuid": "68749650",
      "name": "Some Game",
      "description: "A beautiful game"
    },
    {
      "uuid": "8645645",
      "name": "Amaze Balls",
      "description: "The best game ever"
    }
  ]
}
```

### Developer Facing

#### Promoting an entity to become a developer

##### Request

**POST** ``/developers/:UUID:``

###### Parameters

- **UUID** [REQUIRED]: The UUID of the entity that should become a developer

###### Body

Empty.

##### Response

###### Body

Empty.

#### Demoting an entity from being a developer

Doing so will also remove the entity as a developer from from all games he currently develops.

##### Request

**DELETE** ``/developers/:UUID:``

###### Parameters

- **UUID** [REQUIRED]: The UUID of the entity that should be demoted from being a developer

###### Body

Empty.

##### Response

###### Body

Empty.

#### List games of a developer

**GET** ``/developers/:UUID:/games``

###### Parameters

- **UUID** [REQUIRED]: The UUID of the developer you want to list the games from

###### Body

Empty.

##### Response

###### Body

JSON encoded object with the game's UUIDs as the key and the game information as the value. See ``Change a game's configuration`` response section for an example payload for each game. General structure looks like this:

```javascript
{
  "some-uuid": {
    "uuid": "some-uuid",
    "name": "Some Game",
    …
  },
  "some-other-uuid": {
    …
  }
}
```

#### Adding a game

##### Request

**POST** ``/games``

###### Body

JSON encoded options hash.

- **name** [REQUIRED]: Name of the game
- **description** [REQUIRED]: Text describing the game
- **category** [REQUIRED]: Category of the game
- **configuration** [REQUIRED]: Hash that describes the game. Must include the key ``type`` which can be either ``html5`` or ``flash``.
- **developers** [REQUIRED]: Array of UUIDs of the developers of this game
- **screenshots** [OPTIONAL]: Array of image URLs of screenshots of the game
- **developer_configuration** [OPTIONAL]: Hash of arbitrary data you want to store along with the game
- **venues** [OPTIONAL]: Hash of venues (see [venues](#venues)). The
  key is a venue's name the value is a hash which at least contains
the key ``enabled`` which can be true or false. A not present
key is treated as a disabled venue.


##### Response

###### Body

JSON encoded hash containing the UUID of the newly created game. Example: ``{"uuid": "eb798e00-cb53-012f-4392-58b035f5cdfb"}``

#### Adding a developer to a game

##### Request

**POST** ``/games/:GAME-UUID:/developers/:DEVELOPER-UUID:``

###### Parameters

- **GAME-UUID** [REQUIRED]: The UUID of the game you want to add the developer to
- **DEVELOPER-UUID** [REQUIRED]: The UUID of the developer you want to add to
  the game

###### Body

Empty.

##### Response

###### Body

JSON encoded hash of the game's configuration. See ``Changing a game``
for an example.

#### Removing a developer from a game

##### Request

**DELETE** ``/games/:GAME-UUID:/developers/:DEVELOPER-UUID:``

###### Parameters

- **GAME-UUID** [REQUIRED]: The UUID of the game you want to add the developer to
- **DEVELOPER-UUID** [REQUIRED]: The UUID of the developer you want to add to
  the game

###### Body

Empty.

##### Response

###### Body

JSON encoded hash of the game's configuration. See ``Changing a game``
for an example.

#### Deleting a game

Deleting a game will also remove all developers from this game (relations from developers to the game will be removed, not the developers themselves).

##### Request

**DELETE** ``/games/:UUID:``

###### Parameters

- **UUID** [REQUIRED]: The UUID of the game to be deleted

###### Body

Empty.

##### Response

###### Body

Empty.

#### Change a game's configuration

##### Request

**PUT** ``/games/:UUID:``

###### Parameters

- **UUID** [REQUIRED]: The UUID of the game you want to change

###### Body

JSON encoded options hash.

- **name** [OPTIONAL]: Name of the game
- **description** [OPTIONAL]: Text describing the game
- **category** [OPTIONAL]: Category of a game
- **configuration** [REQUIRED]: Hash that describes the game. See [Game
  Types](#game-types) for more information.
- **developers** [OPTIONAL]: Array of UUIDs of the developers of this game
- **screenshots** [OPTIONAL]: Array of image URLs of screenshots of the game
- **developer_configuration** [OPTIONAL]: Hash of arbitrary data you want to store along with the game
- **venues** [OPTIONAL]: Hash of venues (see [venues](#venues)). The
  key is a venue's name the value  is a hash which at least contains the key ``enabled`` which can be true or false. A not present key is treated as a disabled venue.

Every option that is not present at all in the request will remain unchanged.

##### Response

###### Body

JSON encoded hash of the game's configuration. Example:

```javascript
{
  "name": "Nonsense's Tale",
  "description": "Rumble your way to the treasures of Cpt. McDoodle and his crew. Fight parrots, barrels and pirates!",
  "category": "Jump n Run",
  "configuration": {"type": "html5", "url": "http://example.com/game", "sizes": [{"width": 600, "height": 400}], "fluid-size": true},
  "developers": ["eb6d96b0-cb55-012f-4393-58b035f5cdfb", "ebc08260-cb55-012f-4394-58b035f5cdfb"],
  "screenshots": ["http://example.com/nonsense/1.jpg", "http://example.com/nonsense/2.jpg"],
  "developer_configuration": {"background": "red", "musicTheme": "caribbean"},
  "venues": {"facebook": {"enabled": true}, "spiral-galaxy": {"enabled": false}}
}
```
Note that no matter which options you passed in through the request, the response will always contain the whole configuration.

#### Retrieve a game's configuration

##### Request

**GET** ``/games/:UUID:``

###### Parameters

- **UUID** [REQUIRED]: The UUID of the game you want to retrieve the configuration for

###### Body

Empty.

##### Response

###### Body

JSON encoded hash of the game's configuration. See ``Change a game's configuration`` response section for an example payload.

#### Register a paid subscription to a game

This registers a paid subscription to a game. It uses a token generated by [Stripe.js](https://stripe.com/docs/stripe.js) on our website to charge a credit card.

##### Request

**POST*** ``/games/:UUID:/subscription``

###### Parameters

- **UUID** [REQUIRED]: The UUID of the game you want to subscripe to

###### Body

JSON encoded hash.

- **token** [REQUIRED]: Stripe token to charge the card

##### Response

Status code 201 if subscription was established. 200 if the subscription already existed and 402 if a payment error occured.

#### Cancel a paid subscription to a game

This cancels a paid subscription to a game.

##### Request

**DELETE*** ``/games/:UUID:/subscription``

###### Parameters

- **UUID** [REQUIRED]: The UUID of the game you want to subscripe to

###### Body

Empty

##### Response

Status code 200 if subscription was canceled. 404 if the game has no subscription.

#### Game insights

Get a bunch of statistics for a game

##### Request

**GET** ``/games/:UUID:/insights``

###### Parameters

- **UUID** [REQUIRED]: The UUID of the game you want to subscripe to


##### Response

A JSON object like this:

```javascript
{
  "some-uuid": {
    "players": {
      "overall": 2342,
      "facebook": 1001,
      "spiral-galaxy": 46,
      "embedded": 100
    },
    "impressions": {
      "overall": {
        "logged_in": {
          total: 1031,
          today: 24,
          week: 35,
          month: 89,
          rolling_30_days: [34, 13, 10, 11, 25, 35, …]
        },
        "anonymous": {
          total: 342424,
          today: 325,
          week: 2344,
          month: 7455,
          year: 10433,
          rolling_30_days: [430, 324, 110, 111, 315, 370, …]
        }
      },
      "spiral-galaxy": {
        "logged_in": {
          total: 301,
          today: 5,
          week: 10,
          month: 50,
          rolling_30_days: [14, 17, 10, 21, 5, 3, …]
        },
        "anonymous": {
          total: 4563,
          today: 79,
          week: 150,
          month: 353,
          year: 2546,
          rolling_30_days: [280, 100, 123, 461, 18, 210, …]
        }
      },
      "facebook": {
        …
      },
      "embedded": {
        …
      }
    }
  }
}
```


## Additional Info

### Game Types

Currently supported game types are

#### Initial

This type can only be used for newly created games. It's is **not**
allowed to switch a game from any other game type to ``initial``.

Configuration example:

```javascript
{"type": "initial"}
```

The ``type`` key is mandatory.

#### HTML5

Configuration example:

```javascript
{"type": "html5", "url": "http://example.com/game"}
```

The ``type`` and ``url`` keys are mandatory.

#### Flash

Configuration example:

```javascript
{"type": "flash", "url": "http://example.com/game.swf"}
```

The ``type`` and ``url`` keys are mandatory. ``url`` is pointing to a
SWF flash file.

### Venues

Currently available venues.

#### Facebook

**Key**: ``facebook``

##### Options

#### Spiral Galaxy

**Key**: ``spiral-galaxy``

A Quarter Spiral game portal.


#### Embedded

**Key**: ``embedded``

A white-label venue to embed your game on any website.

Comes with a ``code`` value in the computed configuration that has the complete embed code you need to embed in your website.


### General Configuration

The ``configuration`` object also contains the following properties:

### ``sizes`` [OPTIONAL]

An array that indicates the allowed sizes for the game. Each object in the array must have a ``width`` and ``height`` property. For games using the fluid canvas only the first element of the array is used as a minimal width and height.

#### Examples

```javascript
"configuration": {…, "sizes": [{"width": 600, "height": 400}, {"width": 320, "height": 200}]}
```

### ``fluid-size`` [OPTIONAL]

Boolean property that indicates if a fluid canvas is used. If this property is ``falsy`` a canvas with a fixed size will be used.

#### Examples

```javascript
"configuration": {…, "fluid-size": true}
```

CHANGELOG.md
# 0.0.56 / 2013-08-12

* Fixed sentry app id

# 0.0.55 / 2013-08-12

* Adds getsentry.com exception tracking
* Adds request id tracker
* Updates client dependencies

# 0.0.54 / 2013-05-23

* Makes it possible to fake HTTP methods through a header
* Makes it possible to pass the OAuth token as a query param

# 0.0.53 / 2013-05-13

* Adds ``crossdomain.xml``

# 0.0.52 / 2013-04-07

* Fixes bug when saving games

# 0.0.51 / 2013-04-07

* Adds game insights endpoint

# 0.0.50 / 2013-04-05

* Adds paid game subscriptions and API endpoints
* Fixes bug when saving games with an unchanged developers list

# 0.0.49 / 2013-04-02

* Adds authorization

# 0.0.48

* Respects height margin in embed code's sizing

# 0.0.47

* Reflects games size setting in their embed code

# 0.0.46

* Adds ``uuid`` gem as a dependency to run on Heroku

# 0.0.45

* Updates dependencies

# 0.0.44

* Adds dimension information to games

# 0.0.43

* Updates grape_newrelic

# 0.0.42

* Updates rack
* Updates json

# 0.0.41

* Adds caching to public games list

# 0.0.40

* Adds credit urls for games

# 0.0.39

* Makes it possible to save a credits line with each game

# 0.0.38

* Fixes bug when batch retrieving non existent games

# 0.0.37

* Fixes bug that allowed games to be created without a category

# 0.0.36

* Performance improvements for the public games list

# 0.0.35

* Fixes rolling migration of game categories

# 0.0.34

* Adds categories to games

# 0.0.33

* Migrates game retrieval of developers to new batch API

# 0.0.32

* Adds embed code to public game information

# 0.0.31

* Fixes CORS headers

# 0.0.30

* Adds embed codes to the embedded venue's computed config

# 0.0.29

* Adds embedded venue

# 0.0.28

* Improve Newrelic instrumenting

# 0.0.27

* Adds Newrelic monitoring and ping middleware

# 0.0.26

* Adds venue information to public game info

# 0.0.25

* Fixes bug so that rolling migrated game secrets are persisted

# 0.0.24

* Adds game secrets

# 0.0.23

* Makes it possible to retrieve public game data only for the games specified

# 0.0.22

* Forces new graph-client version

# 0.0.21

* Bumps auth-backend and graph-backend for tests

# 0.0.20

* Bumps datastore dependencies to adopt to the new datastore API

# 0.0.19

* Adds screenshots to the publicly available game information
* Refactors galaxy-spiral to spiral-galaxy in the README

# 0.0.18

* Renames galaxy-spiral to spiral-galaxy
* Gives more intel on error messages for invalid games

# 0.0.17

* Adds an endpoint to retrieve public information about all games

# 0.0.16

* Eases the internal dependencies

# 0.0.15

* Relaxes dependency on auth-client

# 0.0.14

* OPTIONS requests don't need to be authenticated anymore

# 0.0.13

* Makes it possible to initialize games without touching the graph

# 0.0.12

* Adds thin to the main bundle to run on Heroku

# 0.0.11

* Adds special error when a retireved data set is not a game
* Hardens game creation (passing developers has been picky)

# 0.0.10

* Refactors errors to an explicit module
* Moves finder method from API to the game model

# 0.0.9

* Bumps the ``datastore-client``

# 0.0.8

* Bumps the ``graph-client``

# 0.0.7

* Adds the ``initial`` game type
* Adds venues

# 0.0.6

* Removes service-client as an explicit dependency as it is now
  automatically resolved through the datastore-client and graph-client
dependencies.

# 0.0.5

* Moves private gem server credentials to the Gemfile to make it
  possible to deploy on heroku

# 0.0.4

* Moves gem dependencies to the private gem server

# 0.0.3

* Bumps the datastore version

# 0.0.2

* Adds special endpoints to add/remove developers from games

# 0.0.1

The start.


lib/devcenter-backend/game.rb
require 'uuid'

module Devcenter::Backend
  class Game
    MASS_ASSIGNABLE_ATTRIBUTES = [:name, :description, :screenshots, :configuration, :developer_configuration, :venues, :category, :credits, :credits_url]

    attr_accessor :uuid, :name, :description, :category, :credits, :end_of_subscription, :subscription_type, :subscription_customer_id
    attr_writer :screenshots, :developer_configuration
    attr_reader :original_attributes, :token, :secret, :credits_url

    def self.create(token, params)
      developers = params.delete :developers

      raise Error::BaseError.new("Game cannot be created without a category") unless params[:category]

      game = new(token, params.merge(new_game: true))

      ensure_enough_developers!(developers)
      ensure_game_is_valid!(game)

      game.uuid = UUID.new.generate
      connection.datastore.set(game.uuid, token, {'game' => game.to_hash(no_graph: true)})

      unless game.adjust_developers(developers)
        game.destroy
        raise Error::BaseError.new("Can't create game with this developer list!")
      end
      connection.graph.add_role(game.uuid, token, 'game')
      game.mark_as_saved!
      game
    end

    def initialize(token, params = {})
      @token = token
      params = params.clone
      @new_game = params.delete(:new_game)
      params.reject! {|k,v| [:developers, :subscription, :subscription_phasing_out].include?(k.to_sym)}

      raw_update_from_hash(params)

      #TODO: Remove rolling migration
      self.category = 'None' unless category

      generate_secret! unless secret

      @original_attributes = to_hash(no_graph: true)
    end

    def self.find(uuid, token)
      data = connection.datastore.get(uuid, token)

      raise Error::NotFoundError.new("Game #{uuid} not found!") if !data || data.empty?
      raise Error::BaseError.new("Entity not a game (#{uuid})") unless data['game']

      game = new(token, data['game'])
      game.uuid = uuid
      game
    end

    def self.find_batch(uuids, token)
      games = connection.datastore.get(uuids, token)
      return {} unless games
      Hash[games.map do |uuid, data|
        next unless data['game']
        game = new(token, data['game'])
        game.uuid = uuid
        [uuid, game]
      end.compact]
    end

    def self.all(token, options = {})
      game_uuids = options[:uuids] || connection.graph.uuids_by_role(token, 'game')

      find_batch(game_uuids, token).values
    end

    def destroy
      connection = self.class.connection
      connection.datastore.set(uuid, token, {})
      connection.graph.delete_entity(uuid, token)
    end

    def to_hash(options = {})
      hash = {uuid: uuid, name: name, description: description, secret: secret, configuration: configuration, screenshots: screenshots, developer_configuration: developer_configuration, category: category, credits: credits, credits_url: credits_url, subscription: has_subscription?, subscription_phasing_out: has_subscription? && end_of_subscription}

      hash[:developers] = developers unless options[:no_graph]
      if options[:no_graph]
        hash.merge!(
            venues: venues,
            end_of_subscription: end_of_subscription,
            subscription_type: subscription_type,
            subscription_customer_id: subscription_customer_id
        )
      else

        hash[:venues] = venues_with_computed_config
      end

      hash
    end

    def public_information
      ready_venues = venues_with_computed_config.select {|venue, config| config['enabled'] && config['computed']['ready']}.map {|venue, config| venue}
      info = {'uuid' => uuid, 'name' => name, 'description' => description, 'screenshots' => screenshots, 'venues' => ready_venues, 'category' => category, 'credits' => credits, 'credits_url' => credits_url, 'subscription' => has_subscription?}

      if venues_with_computed_config['embedded'] && venues_with_computed_config['embedded']['computed'] && venues_with_computed_config['embedded']['computed']['code']
        info['embed'] = venues_with_computed_config['embedded']['computed']['code']
      end
      info
    end

    def adjust_developers(new_developers)
      old_developers = developers.clone
      developers_to_create = new_developers - old_developers
      developers_to_delete = old_developers - new_developers

      developers_to_create.each do |developer|
        add_developer(developer)
      end
      developers_to_delete.each do |developer|
        remove_developer(developer)
      end
    rescue Service::Client::ServiceError => e
      adjust_developers(old_developers) and return false if e.error =~ /^Relation:.*is invalid!$/
      raise e
    end

    def update_from_hash(hash)
      hash = hash.reject {|k, v| !MASS_ASSIGNABLE_ATTRIBUTES.include?(k.to_sym)}
      raw_update_from_hash(hash)
    end

    def add_developer(developer)
      self.class.connection.graph.add_relationship(developer, uuid, token, 'develops')
    end

    def remove_developer(developer)
      self.class.connection.graph.remove_relationship(developer, uuid, token, 'develops')
    end

    def valid?
      raise Error::ValidationError.new("Games must have a name!") unless name.to_s !~ /^\s*$/
      raise Error::ValidationError.new("Game must have a description!") unless description.to_s !~ /^\s*$/
      raise Error::ValidationError.new("Game must have a category!") unless category.to_s !~ /^\s*$/
      raise Error::ValidationError.new("Game configuration invalid!") unless GameType.valid?(self)
      raise Error::ValidationError.new("Game's credits URL must be a http or https URL") if credits_url && credits_url !~ /^http(s|):\/\//

      Venue.validate_game(self)
    end

    def new_game?
      @new_game
    end

    def mark_as_saved!
      @new_game = false
    end

    def save
      self.class.ensure_game_is_valid!(self)
      self.class.connection.cache.set('last_game_save', Time.now.to_i)
      self.class.connection.datastore.set(uuid, token, {'game' => to_hash(no_graph: true)})
    end

    def developers
      return [] unless uuid
      self.class.connection.graph.list_related_entities(uuid, token, 'develops', direction: 'incoming')
    end

    def configuration=(new_configuration)
      new_configuration = santitize_sizes(new_configuration)
      @configuration = new_configuration
    end

    def configuration
      configuration = @configuration || {}
      defaults = {"sizes" => [{"width" => 600, "height" => 400}]}
      defaults.merge(configuration)
    end

    def developer_configuration
      @developer_configuration || {}
    end

    def screenshots
      @screenshots || []
    end

    def venues
      @venues ||= {}
    end

    def venues_with_computed_config
      Hash[@venues.map do |name, config|
        venue = Venue.venue_for(name, config, self)
        [name, venue.computed_config]
      end]
    end

    def venues=(new_venues)
      new_venues.each do |venue, config|
        venues[venue] = config
      end
    end

    def credits_url=(new_credits_url)
      if new_credits_url =~ /^\s*$/
        @credits_url = nil
      else
        @credits_url = new_credits_url
      end
    end

    def has_subscription?
      !!(subscription_type && (!end_of_subscription || end_of_subscription.to_i > Time.now.to_i) && !(production? && test_subscription?))
    end

    protected
    def production?
      !['development', 'test'].include?(ENV['RACK_ENV'])
    end

    def test_subscription?
      subscription_type != 'live'
    end

    def santitize_sizes(configuration)
      configuration = configuration.clone

      configuration.delete 'sizes' if sizes_blank?(configuration)
      configuration['sizes'].select! {|size| size_valid?(size)} if configuration['sizes']

      configuration
    end

    def sizes_blank?(configuration)
      configuration.keys.include?('sizes') && (!configuration['sizes'] || !configuration['sizes'].kind_of?(Array) || configuration['sizes'].empty?)
    end

    def size_valid?(size)
      size['width'] && size['height'] && size['width'].kind_of?(Numeric) && size['height'].kind_of?(Numeric) && size['width'] >= 0 && size['height'] >= 0
    end

    def raw_update_from_hash(hash)
      hash.each do |key, value|
        value = value.to_hash if value.kind_of?(Hash)
        self.send("#{key}=", value)
      end
    end

    def generate_secret!
      @secret = File.read('/dev/urandom', 32).bytes.map {|e| (e.ord % 64 + 63).chr}.join('')
    end

    def secret=(secret)
      @secret = secret
    end

    def self.connection
      @connection ||= Connection.create
    end

    def self.ensure_enough_developers!(developers)
      raise Error::BaseError.new("Games must have at least one developer!") if !developers || developers.empty?
    end

    def self.ensure_game_is_valid!(game)
      # game validation raises errors from within itself
      game.valid?
    end
  end
end
